query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java

```diff
@@ -323,34 +324,31 @@ private String convertExpressionToMapValueColumn(Expression expression) {
     throw new IllegalArgumentException("operator supports multi value column only");
   }
 
-  private LiteralConstant[] convertExpressionToMapLiterals(Expression expression) {
-    LiteralConstant[] literals = new LiteralConstant[2];
+  private List<LiteralConstant> convertExpressionToMapLiterals(Expression expression) {
+    List<String> literals = new ArrayList<>(List.of("", ""));
```

Comment:
If we're going to clean up the implementation, let's do it right and return an object rather than a list.
```
@Value
class MapKeyValueArguments {
  LiteralConstant key;
  Optional<LiteralConstant> value;
}
```

Reply:
can `Map.Entry<LiteralConstant, Optional<LiteralConstant>>` be used as tuple?

Reply:
I'm a bit on the fence. I probably wouldn't do it, but I wouldn't call it out on review. In general, we'd rather create a new class than use an Entry or Pair class, but in this case it's reasonable to use Entry since we're representing a key and value (Pair pretty much never makes sense, because left and right have no meaning). The reason I'd still lean against it is that we're kind of mixing a structural paradigm with an object-oriented one. What we're producing here "is a" map entry, but it would never be used in a map - we're just using an existing unrelated class because it has the structure we're looking for.

Reply:
If the key is an index like 0 or 1, then why do we need it to be a literalConstant ? Then at the places where we access using 0/1, we would need to create a literalConstant first.

Reply:
Currently, we use a list. Here we do `x.get(map_key_index) where map_key_index is 0`. 
If we go by the suggested approach, then we would have to create a literal constant for 0/1. Since the key is fairly simple (only 0/1), can we use something like this ?
```
@Value
class MapKeyValueArguments {
  int key;
  Optional<LiteralConstant> value;
}
```

Reply:
As a second thought (after looking at the test of the LIKE operator with attribute expression), should we handle attribute expression for `CONTAINS_KEY` and `NOT_CONTAINS_KEY` as part of this PR? 

Is there any other place, we have not still handled `attribute expression`?

Reply:
Everything is covered now I guess.

Reply:
So, as part of this PR, shall we cover `CONTAINS_KEY and NOT_CONTAINS_KEY` for attribute expression as well?

Reply:
> Currently, we use a list. Here we do x.get(map_key_index) where map_key_index is 0.
If we go by the suggested approach, then we would have to create a literal constant for 0/1. Since the key is fairly simple (only 0/1), can we use something like this ?

How does the `int key` value be used? How will we get LiteralConsant, will we still have List<LiteralConsant>? Based on aaron's suggestion, you can do something as below. And, you can access `key` literal constant directly from the object.

```
diff --git a/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/MapKeyValueArguments.java b/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/MapKeyValueArguments.java
index 84d627a..89ecba1 100644
--- a/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/MapKeyValueArguments.java
+++ b/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/MapKeyValueArguments.java
@@ -1,2 +1,11 @@
-package org.hypertrace.core.query.service.pinot;public class MapKeyValueArguments {
+package org.hypertrace.core.query.service.pinot;
+
+import java.util.Optional;
+import lombok.Value;
+import org.hypertrace.core.query.service.api.LiteralConstant;
+
+@Value
+class MapKeyValueArguments {
+  LiteralConstant key;
+  Optional<LiteralConstant> value;
 }
diff --git a/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java b/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java
index 557a670..e99f1eb 100644
--- a/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java
+++ b/query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java
@@ -9,10 +9,10 @@ import static org.hypertrace.core.query.service.api.Expression.ValueCase.LITERAL
 import com.google.common.base.Joiner;
 import com.google.protobuf.ByteString;
 import java.util.AbstractMap.SimpleEntry;
-import java.util.ArrayList;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map.Entry;
+import java.util.Optional;
 import org.hypertrace.core.query.service.ExecutionContext;
 import org.hypertrace.core.query.service.api.Expression;
 import org.hypertrace.core.query.service.api.Filter;
@@ -156,38 +156,41 @@ class QueryRequestToPinotSQLConverter {
           builder.append(")");
           break;
         case NOT_CONTAINS_KEY:
-          List<LiteralConstant> kvp = convertExpressionToMapLiterals(filter.getRhs());
+          MapKeyValueArguments mapKeyValueArguments =
+              convertExpressionToMapLiterals(filter.getRhs());
           builder.append(convertExpressionToMapKeyColumn(filter.getLhs()));
           builder.append(" != ");
-          builder.append(convertLiteralToString(kvp.get(MAP_KEY_INDEX), paramsBuilder));
+          builder.append(convertLiteralToString(mapKeyValueArguments.getKey(), paramsBuilder));
           break;
         case CONTAINS_KEY:
-          kvp = convertExpressionToMapLiterals(filter.getRhs());
+          mapKeyValueArguments = convertExpressionToMapLiterals(filter.getRhs());
           builder.append(convertExpressionToMapKeyColumn(filter.getLhs()));
           builder.append(" = ");
-          builder.append(convertLiteralToString(kvp.get(MAP_KEY_INDEX), paramsBuilder));
+          builder.append(convertLiteralToString(mapKeyValueArguments.getKey(), paramsBuilder));
           break;
         case CONTAINS_KEYVALUE:
-          kvp = convertExpressionToMapLiterals(filter.getRhs());
+          mapKeyValueArguments = convertExpressionToMapLiterals(filter.getRhs());
           String keyCol = convertExpressionToMapKeyColumn(filter.getLhs());
           String valCol = convertExpressionToMapValueColumn(filter.getLhs());
           builder.append(keyCol);
           builder.append(" = ");
-          builder.append(convertLiteralToString(kvp.get(MAP_KEY_INDEX), paramsBuilder));
+          builder.append(convertLiteralToString(mapKeyValueArguments.getKey(), paramsBuilder));
           builder.append(" AND ");
           builder.append(valCol);
           builder.append(" = ");
-          builder.append(convertLiteralToString(kvp.get(MAP_VALUE_INDEX), paramsBuilder));
+          builder.append(
+              convertLiteralToString(mapKeyValueArguments.getValue().get(), paramsBuilder));
           builder.append(" AND ");
           builder.append(MAP_VALUE);
           builder.append("(");
           builder.append(keyCol);
           builder.append(",");
-          builder.append(convertLiteralToString(kvp.get(MAP_KEY_INDEX), paramsBuilder));
+          builder.append(convertLiteralToString(mapKeyValueArguments.getKey(), paramsBuilder));
           builder.append(",");
           builder.append(valCol);
           builder.append(") = ");
-          builder.append(convertLiteralToString(kvp.get(MAP_VALUE_INDEX), paramsBuilder));
+          builder.append(
+              convertLiteralToString(mapKeyValueArguments.getValue().get(), paramsBuilder));
           break;
         default:
           rhs = handleValueConversionForLiteralExpression(filter.getLhs(), filter.getRhs());
@@ -331,31 +334,37 @@ class QueryRequestToPinotSQLConverter {
     throw new IllegalArgumentException("operator supports multi value column only");
   }
 
-  private List<LiteralConstant> convertExpressionToMapLiterals(Expression expression) {
-    List<String> literals = new ArrayList<>(List.of("", ""));
+  private MapKeyValueArguments convertExpressionToMapLiterals(Expression expression) {
+    LiteralConstant mapKey = null;
+    Optional<LiteralConstant> mapValue = Optional.empty();
+
     if (expression.getValueCase() == LITERAL) {
       LiteralConstant value = expression.getLiteral();
       if (value.getValue().getValueType() == ValueType.STRING_ARRAY
           && value.getValue().getStringArrayCount() == 2) {
-        literals.set(0, value.getValue().getStringArray(0));
-        literals.set(1, value.getValue().getStringArray(1));
+        mapKey =
+            LiteralConstant.newBuilder()
+                .setValue(Value.newBuilder().setString(value.getValue().getStringArray(0)).build())
+                .build();
+
+        mapValue =
+            Optional.of(
+                LiteralConstant.newBuilder()
+                    .setValue(
+                        Value.newBuilder().setString(value.getValue().getStringArray(1)).build())
+                    .build());
       } else if (value.getValue().getValueType() == ValueType.STRING) {
-        literals.set(0, value.getValue().getString());
+        mapKey =
+            LiteralConstant.newBuilder()
+                .setValue(Value.newBuilder().setString(value.getValue().getString()).build())
+                .build();
       } else {
         throw new IllegalArgumentException(
             "Unsupported arguments for CONTAINS_KEY / CONTAINS_KEYVALUE operator");
       }
     }
 
-    List<LiteralConstant> literalConstantList = new ArrayList<>();
-    for (int i = 0; i < 2; i++) {
-      literalConstantList.add(
-          LiteralConstant.newBuilder()
-              .setValue(Value.newBuilder().setString(literals.get(i)).build())
-              .build());
-    }
-
-    return literalConstantList;
+    return new MapKeyValueArguments(mapKey, mapValue);
   }
 
   /** TODO:Handle all types */

```

Reply:
>So, as part of this PR, shall we cover CONTAINS_KEY and NOT_CONTAINS_KEY for attribute expression as well? Like in this case convertExpressionToMapLiterals which returns key in index 0. For attribute expression, it needs to fetch from its subPath, right?

Then also it will still have rhs right ?
Like we can say `Span.tags CONTAINS_KEY 'flags'` or `lhs CONTAINS_KEY rhs`
For attribute expression how will we use the rhs expression ?

Currently we have these function : 
```
  public static Filter createContainsKeyFilter(AttributeExpression complexAttributeExpression) {
    return createContainsKeyFilter(
        complexAttributeExpression.getAttributeId(), complexAttributeExpression.getSubpath());
  }

  public static Filter createContainsKeyFilter(String column, String value) {
    return Filter.newBuilder()
        .setOperator(Operator.CONTAINS_KEY)
        .setLhs(createSimpleAttributeExpression(column))
        .setRhs(createStringLiteralValueExpression(value))
        .build();
  }
```

Reply:
>How does the int key value be used? How will we get LiteralConsant, will we still have List? Based on aaron's suggestion, you can do something as below. And, you can access key literal constant directly from the object. e.g builder.append(convertLiteralToString(mapKeyValueArguments.getKey(), paramsBuilder));

Ok got it now. Earlier I thought 0 is key. But you are proposing that key should be `Span.tags` and value should be say `span.kind`. 

Should I make this change then ?

Reply:
> Then also it will still have rhs right ?
Like we can say Span.tags CONTAINS_KEY 'flags' or lhs CONTAINS_KEY rhs

Yes, you are right. Rhs is a key in map. I got confused.

Reply:
>> How does the int key value be used? How will we get LiteralConsant, will we still have List? Based on aaron's suggestion, you can do something as below. And, you can access key literal constant directly from the object. e.g builder.append(convertLiteralToString(mapKeyValueArguments.getKey(), paramsBuilder));

> Ok got it now. Earlier I thought 0 is key. But you are proposing that key should be Span.tags and value should be say span.kind.

For a map, there are two arguments one is key and the other is value - e.g x['key'] = 'value'
So, here `Span.tags` is a map, the `span.kind` is a key arg. 
In the case of `CONTAINS_KEY` and `NOT_CONTAINS_KEY`, it will have only one `key` arg. But, in the case of `CONTAINS_KEY_VALUE`, it will have both `key` and `value` args, and so `value` arg is optional in structure.
e.g `Span.tags['span.kind'] = 'client'`, for this `key` will be `span.kind` and `value` will be `client`

> Should I make this change then ?
Yes, we can do that.

Reply:
Alternatively, we can wait then drop this whole thing. Assuming we deprecate CONTAINS_KEY_VALUE and remove the array around the contains_key arg, this whole function can just go away I think?

Reply:
Yes. We can simplify that and move inline maybe.

Reply:
resolved